server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -78,132 +94,336 @@ public NodeConnectionsService(Settings settings, ThreadPool threadPool, Transpor
+     * or vice versa. The connection/disconnection process continues until all listeners have been removed, at which point it becomes idle
+     * again.
+     * <p>
+     * Additionally if the last step of the process was a disconnection then this target is removed from the current set of targets. Thus
+     * if this {@link ConnectionTarget} is idle and in the current set of targets then it should be connected.
+     * <p>
+     * All of the {@code listeners} are awaiting the completion of the same activity, which is either a connection or a disconnection.  If
+     * we are currently connecting and then {@link ConnectionTarget#disconnect()} is called then all connection listeners are
+     * removed from the list so they can be notified of failure; once the connecting process has finished a disconnection will be started.
+     * Similarly if we are currently disconnecting and then {@link ConnectionTarget#connect(ActionListener)} is called then all
+     * disconnection listeners are immediately removed for failure notification and a connection is started once the disconnection is
+     * complete.
+     */
+    private class ConnectionTarget {
+        private final DiscoveryNode discoveryNode;
+
+        private PlainListenableActionFuture<Void> future = PlainListenableActionFuture.newListenableFuture();
+        private ActivityType activityType = ActivityType.IDLE; // indicates what any listeners are awaiting
+
+        private final AtomicInteger consecutiveFailureCount = new AtomicInteger();
+
+        private final Runnable connectActivity = () -> threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(new AbstractRunnable() {
+            @Override
+            protected void doRun() {
+                assert Thread.holdsLock(mutex) == false : "mutex unexpectedly held";
+                transportService.connectToNode(discoveryNode);
+                consecutiveFailureCount.set(0);
+                logger.debug("connected to {}", discoveryNode);
+                onCompletion(ActivityType.CONNECTING, null, disconnectActivity);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                assert Thread.holdsLock(mutex) == false : "mutex unexpectedly held";
+                final int currentFailureCount = consecutiveFailureCount.incrementAndGet();
+                // only warn every 6th failure
+                final Level level = currentFailureCount % 6 == 1 ? Level.WARN : Level.DEBUG;
+                logger.log(level, new ParameterizedMessage("failed to connect to {} (tried [{}] times)",
+                    discoveryNode, currentFailureCount), e);
+                onCompletion(ActivityType.CONNECTING, e, disconnectActivity);
+            }
+
+            @Override
+            public String toString() {
+                return "connect to " + discoveryNode;
+            }
+        });
+
+        private final Runnable disconnectActivity = new AbstractRunnable() {
```

Comment:
I contemplated putting this on a thread pool, but AFAIK disconnection is itself a fast operation. We do not wait for disconnections to complete _in general_ because there may be a (slow) connection attempt already in progress, and the disconnection is then enqueued behind the connection, but if that isn't the case then I think it's ok to disconnect synchronously like this.

Reply:
@DaveCTurner can you please comment to `disconnectActivity` why you don't use a thread pool here?